core/base/batch_instantiation.hpp

```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_BASE_BATCH_INSTANTIATION_HPP_
+#define GKO_PUBLIC_CORE_BASE_BATCH_INSTANTIATION_HPP_
+
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/batch_csr.hpp>
+#include <ginkgo/core/matrix/batch_dense.hpp>
+#include <ginkgo/core/matrix/batch_ell.hpp>
+#include <ginkgo/core/matrix/batch_identity.hpp>
+#include <ginkgo/core/preconditioner/batch_jacobi.hpp>
+
+namespace gko {
+namespace batch {
+
+/**
+ * Instantiates a template for each valid combination of value type, batch
+ * matrix type, and batch preconditioner type. This only allows batch matrix
+ * type and preconditioner type also uses the same value type.
+ *
+ * @param _macro  A macro which expands the template instantiation
+ *                (not including the leading `template` specifier).
+ *                Should take three arguments, where the first is replaced by
+ *                the value type, the second by the matrix, and the third by the
+ *                preconditioner.
+ *
+ * @note the second and third arguments only accept the base type.s
+ */
+#define GKO_INSTANTIATE_FOR_BATCH_VALUE_MATRIX_PRECONDITIONER(_macro)          \
```

Comment:
I think this needs to be unified with the changes from #1629, but I don't immediately see how to do that.

Reply:
I reuse the part in types.hpp.

Reply:
I update the macro usage a bit such that make the for-stack easy and clear.
I think it can not reuse the main stack of dispatch type.
one dispatches the device data structure which are only available in `namespace cuda/hip/...`, but the other uses the main class.